### ISSUE TITLE
The concourse_workers_containers Prometheus metric reports values from old, currently non-existent workers

### ISSUE THREAD

**Issue by kmdouglass** [concourse/concourse#4341]
## Bug Report

Four days ago we deployed the release candidate 7c63f45. Immediately after doing so we noticed that the `concourse_workers_containers` Prometheus metric began reporting a number of containers for workers that no longer exist but at one point did. These workers existed at one point because they were part of an AWS autoscaling group (ASG) and eventually were retired and destroyed as part of the natural lifecycle of the ASG.

Prior to this RC, the metric would not report the number of containers for previous workers; it would only report the number of containers for active workers. Now it reports the number of containers for both the current workers and previous but currently non-existent workers.

After four days of running the release candidate in production, we have accumulated a large number of time series for workers that no longer exist. This can be seen in the plot below:

![image](https://user-images.githubusercontent.com/3697676/64123592-1e1c9b80-cda5-11e9-8a3f-25c5e72c48c5.png)

The plot shows the number of time series that belong to the `concourse_workers_containers` metric on the y-axis and the time since ~2019/08/29 on the x-axis. (We deployed the RC around 8:30 AM on the 29'th of August.) The PromQL query is:

```
count(count by (worker)(concourse_workers_containers{worker!~".*mac-.*",worker!="deployer"}))
```

The inner `count by (worker)` essentially returns a vector of 1's, one for each time series returned by `concourse_workers_containers`, and the second count sums all of these 1's. The plot is constant over the weekend because we had no workers running. Once Monday, September 2 came and new workers were spawned, the number of time series began to grow again.

## Steps to Reproduce

1. Deploy RC 7c63f45 into production with a pool of workers that can dynamically grow and shrink over time.
2. Observe that the `concourse_workers_containers` Prometheus metric will continue to report a value for workers that existed at one point but no longer exist.

## Expected Results

I expect that the Prometheus `concourse_workers_containers` metric to report a number of containers for workers that exist at the time the metric was scraped. Old workers that no longer exist should not be included in the metric.

## Actual Results
The `concourse_workers_containers` metric reports both the number of containers for active workers and for workers that existed at one point but no longer exist.

## Additional Context

- The previous RC that we were running that DID NOT have this bug is 5672114.
- The output of `fly workers` is what I would expect; it only shows the workers currently registered with the ATC and not the old ones.
- *Update 2019-09-04* Restarting the ATC actually clears out these false workers from memory so that they will no longer appear in the metrics. We're now restarting the ATC nightly to clear them.

## Version Info

* Concourse version: 5.5.0 RC 7c63f45
* Deployment type: Binary
* Infrastructure/IaaS: AWS with workers in an autoscaling group
* Did this used to work? Yes

**Comment by kmdouglass** [comment]
I've done a bit of research. I think the bug was introduced in commit 5824dfd. In this commit, additional `team` and `tags` labels were added to the metrics for the `worker-volumes` and `worker-containers`.

```
--- a/atc/metric/emitter/prometheus.go
+++ b/atc/metric/emitter/prometheus.go
@@@ -507,30 -504,9 +519,30 @@@ func (emitter *PrometheusEmitter) worke
                return
        }

-       emitter.workerVolumes.WithLabelValues(worker, platform).Set(float64(volumes))
+       emitter.workerVolumes.WithLabelValues(worker, platform, team, tags).Set(float64(volumes))
  }
```

However, the garbage collector for the metrics is still working on the `worker` and `platform` labels alone:

```console
rg DeleteLabelValues
atc/metric/emitter/prometheus.go
653:                            // but DeleteLabelValues requires an exact match on the label set.
656:                                    emitter.workerContainers.DeleteLabelValues(worker, platform)
657:                                    emitter.workerVolumes.DeleteLabelValues(worker, platform)
658:                                    emitter.workerTasks.DeleteLabelValues(worker, platform)
```

So I think that the Prometheus metrics for the containers and volumes are no longer being garbage collected.

**Comment by cirocosta** [comment]
ooh good catch, that makes sense indeed!

Thanks for going through the code and drilling down to the GC of the metrics we observe.

Would you mind submitting a PR for it? The contribution process can be found here: [Contribution process](https://github.com/concourse/concourse/blob/cfe7332347e805e65e32426815747d34e6fcb786/CONTRIBUTING.md#contribution-process).

Thank you very much!

**Comment by kmdouglass** [comment]
@cirocosta Thanks for the confirmation. We have a ticket open on our board to address this; I'll open a PR in the coming days when I get a chance to work on it.

**Comment by cirocosta** [comment]
awesome! if you need any help, just let us know 😁  thanks!

**Comment by kmdouglass** [comment]
@cirocosta @marco-m @wagdav @aledeganopix4d 

I opened a draft PR that proposes a solution to the problem: https://github.com/concourse/concourse/pull/4446

I am now working on testing it. In the meantime, I would appreciate any feedback about this approach, especially from the Concourse team.

**Comment by kmdouglass** [comment]
@cirocosta I opened the PR #4446  for review. Please let me know whether you have questions.

**Comment by pivotal-jamie-klassen** [comment]
Fixed in #4446
